sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncClient.java

```diff
@@ -581,49 +593,56 @@ public GlobalThroughputControlConfigBuilder createGlobalThroughputControlConfigB
         return new GlobalThroughputControlConfigBuilder(this, databaseId, containerId);
     }
 
-    void openConnectionsAndInitCaches() {
-        blockListVoidResponse(openConnectionsAndInitCachesInternal());
-    }
-
     WriteRetryPolicy getNonIdempotentWriteRetryPolicy() {
         return this.nonIdempotentWriteRetryPolicy;
     }
 
-    private Mono<List<Void>> openConnectionsAndInitCachesInternal() {
-        int concurrency = 1;
-        int prefetch = 1;
-        if (this.proactiveContainerInitConfig != null) {
-            return Flux.fromIterable(this.proactiveContainerInitConfig.getCosmosContainerIdentities())
-                    .flatMap(
-                        cosmosContainerIdentity -> Mono.just(this
-                            .getDatabase(containerIdentityAccessor.getDatabaseName(cosmosContainerIdentity))
-                            .getContainer(containerIdentityAccessor.getContainerName(cosmosContainerIdentity))),
-                        concurrency,
-                        prefetch
-                    )
-                    .flatMap(
-                        cosmosAsyncContainer -> cosmosAsyncContainer
-                            .openConnectionsAndInitCaches(
-                                this.proactiveContainerInitConfig.getProactiveConnectionRegionsCount()),
-                        concurrency,
-                        prefetch)
-                    .collectList();
-        }
-        return Mono.empty();
+    void openConnectionsAndInitCaches() {
+        final Duration lastSuccessResponseTimeout = Duration.ofSeconds(5);
+
+        asyncDocumentClient.submitOpenConnectionTasksAndInitCaches(
+                        proactiveContainerInitConfig,
+                        AsyncDocumentClient.OpenConnectionAggressivenessHint.AGGRESSIVE
+                )
+                .subscribeOn(CosmosSchedulers.OPEN_CONNECTIONS_BOUNDED_ELASTIC)
+                .subscribe();
+
+        wrapSourceFluxWithSoftTimeoutAndFallback(
+                proactiveOpenConnectionsProcessor.getOpenConnectionsPublisher().sequential(),
+                Flux.empty(),
+                lastSuccessResponseTimeout,
+                CosmosSchedulers.OPEN_CONNECTIONS_BOUNDED_ELASTIC
+        ).blockLast();
     }
 
-    private void blockListVoidResponse(Mono<List<Void>> voidListMono) {
-        try {
-            voidListMono.block();
-        } catch (Exception ex) {
-            final Throwable throwable = Exceptions.unwrap(ex);
 
-            if (throwable instanceof CosmosException) {
-                throw (CosmosException) throwable;
-            } else {
-                throw Exceptions.propagate(throwable);
-            }
-        }
+    void openConnectionsAndInitCaches(Duration timeout) {
+        asyncDocumentClient.submitOpenConnectionTasksAndInitCaches(
+                        proactiveContainerInitConfig,
+                        AsyncDocumentClient.OpenConnectionAggressivenessHint.AGGRESSIVE
```

Comment:
let us say if when the timeout reached, we are still in the connection tasks submit stage (for example the getAddresses etc takes time), when & how are we going to change the aggressiveness?

Reply:
There is a Flux at the end of this method which blocks for the given duration and re-instantiates the open connections sink, changes the aggressiveness and re-subscribes to the open connections sink.

> we are still in the connection tasks submit stage (for example the getAddresses etc takes time)

Yes this could take time, that's why I thought of not completing the open connections sink. Get address requests could take any amount time therefore the sink should always be in a state to accept open connection tasks sent by which ever class that takes a dependency on this sink - (`GlobalAddressResolver`, `RntbdConnectionStateListener `etc.)